### DISCUSSION TITLE
Graphically see the difference between two versions of a .mm file, for use with git

### DISCUSSION THREAD

**Discussion by cedric-airsupplies** [freeplane/freeplane#1368] [Ideas]
I'm using mindmaps to maintain todo lists for the software I write. I store this .mm file along with the sourcecode in a git repository.
It would be nice if I could graphically see the difference between two versions of the .mm file. Maybe this can be implemented by a command line switch to diff two files like this, so freeplane doesn't need to know  about git:
$ freeplane --diff firstfile.mm secondfile.mm

**Comment by macmarrum** [comment]
This can be achieved with an add-on.

Here's a script that does a basic comparison of the currently open mind map to its earlier version (saved as a different file), i.e. similar to how document comparison works in LibreOffice.

Save [compareMindmap.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/compareMindmap.groovy) to `<user-dir>/scripts` and restart Freeplane.
You'll find it in `Mac2->Compare Mindmap`.

**Note:**
- _You can use `Tools->Open user directory` to get to `<user-dir>`_
- _To have the script appear in `Tools->Scripts`, remove `="/menu_bar/Mac2"` from the script's first line_

Changes are indicated by marking modified nodes with special `diff:` styles. The styles will be created, if missing. You can then save the styles on your original mind map, modify them to your liking and they will be used the next time you'll run the script.

Any feedback is welcome.

**Reply by macmarrum** [reply]
This is how the auto-created styles look like:

![diff-styles](https://github.com/freeplane/freeplane/assets/7197131/f6265f07-6049-4126-898a-093502efcc64)

And their application

![diff-styles-in-use](https://github.com/freeplane/freeplane/assets/7197131/7b5e99b7-9b88-414d-ae4d-c530892c22fb)

Sample mind map: 
[compare-mindmap-new~diff.mm.zip](https://github.com/freeplane/freeplane/files/12489626/compare-mindmap-new.diff.mm.zip)

**Reply by macmarrum** [reply]
[freeplane-remote-diff.py](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/python/freeplane-remote-diff.py) is a Python script that connects to a running instance of Freeplane and instructs it to run `MindMapComparator.compare(oldFile, newFile)`, i.e. to do the graphical comparison:

Example
```bash
$ freeplane-remote-diff.py mindmap-ver-1.mm mindmap.mm
```

It can also be used as a git difftool: 

```bash
$ git config --global --add difftool.freeplane.cmd 'path/to/your/freeplane-remote-diff.py "$LOCAL" "$REMOTE"'
$ git difftool -t freeplane HEAD^ mindmap.mm
```

The script requires the following Groovy files to be present in your Freeplane instalation:
- [remoteControl.groovy](https://github.com/macmarrum/freeplane-scripting/blob/main/src/init/remoteControl.groovy) in `<user-dir>/scripts/init`
- [MindMapComparator.groovy](https://github.com/macmarrum/freeplane-scripting/blob/main/src/lib/MindMapComparator.groovy) in `<user-dir>/lib`

**Comment by jcdouet** [comment]
Hello,

Thank you for the pretty work. I would like to report a problem that prevents me for fully using this machinery in my daily work…
To give it a try, I used the resulting command (compare maps) on two simple maps, it works, it's cool.

Now, I have noticed that if one of the maps has « calendar alerts » within it, the comparison fails each time.
Can you reproduce that behavior ?

Best regards,
   Jean-Charles.

**Reply by macmarrum** [reply]
@jcdouet I'm glad the solution is useful to you.

> I have noticed that if one of the maps has « calendar alerts » within it, the comparison fails each time.

I'm not sure what you mean. Could you please attach your mind maps (the current and an old version) for which the comparison fails and list the steps that lead to the error? You can obfuscate your mind maps, if needed → [Map Obfuscator](https://github.com/macmarrum/freeplane-MapObfuscator)

**Reply by jcdouet** [reply]
Well, after verifying, the problem is not reproducing only for maps with a calendar alert (which I thought at first).
I used the map obfuscator, and comparing directly between obfuscated maps, I got the same error.

For instance, let's say that the first map is « Économie.mm » :
- I obfuscate, and I obtain « Économie~obfuscated.mm »
- Then I cut a few branches intentionnally, saving this new map into « tmp_Économie~obfuscated.mm »
- Then I click onto « compare Mind maps » script, where « tmp_Économie… » is the old map.
- Then an error pops up, for which I have made a screenshot.

Please find in the attached zip 3 files : the old map, the new map and the screenshot.

Once again, sorry, I thought it was only for the « calendar alert » case, but the problem still remains for many maps. Strangely, for other maps (that I cannot remind of), this diff tool used to work correctly. I cannot manage to remember which change was made to freeplane configuration either in the while.
[case_for_diff_pb.zip](https://github.com/freeplane/freeplane/files/13064190/case_for_diff_pb.zip)

**Reply by macmarrum** [reply]
I executed the script on the attached mind maps. It worked OK for me.

Please make sure you use the latest version of the script (re-download it, restart Freeplane) and try again.
- [compareMindmap.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/compareMindmap.groovy) in `<user-dir>/scripts` 
- [MindMapComparator.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/lib/MindMapComparator.groovy) in `<user-dir>/lib`

**Comment by EdoFro** [comment]
> I'm using mindmaps to maintain todo lists for the software I write. I store this .mm file along with the sourcecode in a git repository.

@cedric-airsupplies,
Sorry this a little late, but may be useful.

I created a while ago an addon that helps saving mindmaps without folding and modification dates information. This way it's easier to see the differences when using git comparison.
 https://github.com/EdoFro/Freeplane_Map-Saving-Options
